### ISSUE TITLE
HttpUtility.ParseQueryString does HTML decoding first on CoreFX, but not on NetFX

### ISSUE THREAD

**Issue by JonHanna** [dotnet/corefx#23582]
The version of `HttpUtility.ParseQueryString` in CoreFX does HTML decoding first, differing from desktop:

```C#
HttpUtility.ParseQueryString("name=value&amp;name2=value2").ToString()
// name=value&amp%3bname2=value2 on Desktop
// name=value&name2=value2 on CoreFX

HttpUtility.ParseQueryString("?name=valu&#xe9;")["name"]
// valu on Desktop
// valué on CoreFX
```

#23573 is a knock-on effect of this. (Closing as such).
#23574 is an indirect knock-on effect in that ? is stripped both before and after HTML decoding that shouldn't be done at all.

**Comment by JonHanna** [comment]
Fixing this could break someone who had depended on this behaviour, but the issue itself could break someone depending on compatibility with netfx.

**Comment by davidsh** [comment]
Getting more compatible with NETFX is usually a higher priority.  It's unfortunate that it wasn't fixed before .NET Core 2.0.  But fixing in .NET Core 2.1 to improve its compat w/ NETFX would be a good thing.

cc: @karelz @danmosemsft

**Comment by karelz** [comment]
Does it mean that Desktop never decoded HTML in this API?

Looking at code it should: [Desktop code](http://referencesource.microsoft.com/#System.Web/httpserverutility.cs,1276) vs. [CoreFX code](https://github.com/dotnet/corefx/blob/2117bae3a6c9736033e6d1659821ff3a1069920b/src/System.Web.HttpUtility/src/System/Web/HttpUtility.cs#L39-L115)

**Comment by JonHanna** [comment]
@karelz where? The desktop code line you link to calls the overload that takes an encoding, that in turn does some null checks, strips off any initial `?` and constructs a `HttpValueCollection`. That ctor calls `FillFromString` which has logic much as on corefx but without the HTML decoding. (There's URL decoding in both, which is another thing, and certainly correct).

**Comment by karelz** [comment]
I guess I mixed it up with the URL decoding then.

**Comment by karelz** [comment]
@davidsh can you please run it by AppCompat?
We should also get 2.0.x approval, then we can check it into 2.1. If we don't get 2.0.x servicing approval then it is questionable if we want to fix it in 2.1 IMO.

**Comment by JonHanna** [comment]
*phew* (was a bit worried it was right in front of my eyes and I was somehow not seeing it, because it wouldn't be the first time that ever happened).

**Comment by karelz** [comment]
OK, doing compat discussion here: @ericstj @stephentoub @terrajobst @Petermarcu @danmosemsft  @weshaggard -- this looks like potentially impactful breaking change we introduced in 2.0 against Desktop.

The API is new in .NET Core 2.0: https://apisof.net/catalog/System.Web.HttpUtility.ParseQueryString(String)
It is part of .NET Framework 2.0+ and also part of .NET Standard 2.0.
Telemetry shows 2.25% usage from API Port telemetry (non-trivial). 1.25% usage on nuget.org.

Questions:
* Should we fix it in 2.1 to match Desktop? (area owners recommendation: Yes)
* Should we (proactively) fix it in 2.0.x servicing, or wait for more reports first? (area owners recommendation: 2.0.x servicing)

**Comment by danmosemsft** [comment]
I would fix it in servicing.

**Comment by danmosemsft** [comment]
@karelz who is doing the fix? do we want to get into 2.0.2 - if so need to be quickish.

**Comment by karelz** [comment]
AFAIK the fix is ready, we "just" need to decide if we want to take it from compat perspective - I expected @davidsh to drive it, but I can send email to shiproom given no one replied here :(.

Thanks for catching it @danmosemsft!

**Comment by karelz** [comment]
Reopening to track porting to release/2.0.0 branch.

**Comment by karelz** [comment]
2.0.x branch fixed in PR #23918.
